### ISSUE TITLE
Saving model and loading model on a separate session gives wrong predictions

### ISSUE THREAD

**Issue by AchiraFernando** [tensorflow/tensorflow#27887]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: No
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Windows 10
- **Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device**: No
- **TensorFlow installed from (source or binary)**: using pip install tensorflow-gpu
- **TensorFlow version (use command below)**:1.12.0
- **Python version**: 3.6.3
- **Bazel version (if compiling from source)**: - 
- **GCC/Compiler version (if compiling from source)**: - 
- **CUDA/cuDNN version**: 9
- **GPU model and memory**: GeForce 940MX
- **Exact command to reproduce**: As described below

### Describe the problem

As soon as training is finished, generating predictions on the testing dataset gives an accuracy > 70.

Then I save the model using model.save(...) method.

When Trying to load the model in a separate session, the predictions are around 20s.

I can confirm that i'm using the same testing dataset and same preprocessing steps that I've done when predicting after training.

To get over with the doubt on the different tokenizers, I've saved the tokenizer to a tokenizer.pickle file which will be loaded before conducting texts_to_sequences and pad_sequences on the testing text data (this didn't make any difference).

I've tried compiling the model after loading before making predictions, this didn't work as well.

### Source code / logs

`model.save('model.h5') #for saving the model`

---------------------------------------------------------------------------

`#Predicting after saving on a different session`

`trained_model = load_model('model.h5') #for loading the model in a different session.`

`# load tokenizer`
`tokenizer = Tokenizer()`
`with open('trained_model/tokenizer.pickle', 'rb') as handle:`
`  tokenizer = pickle.load(handle)`

`test_revs = pd.read_csv('test_dataset.csv')`

`test_revs.loc[:, 'rating'] = test_revs['rating'].apply(points_to_class) #converting decimal ratings to integer classes `

`actual_texts = test_revs['text']`
`actual_ratings = test_revs['rating']`

`final_Y_test = to_categorical(actual_ratings, 5)`

`actual_text_tokens = add_doc_to_vocab(actual_texts.tolist()) # preprocessing method`

`sequences_test = tokenizer.texts_to_sequences(actual_text_tokens)`

`X_test = pad_sequences(sequences_test, maxlen=1939, padding='post')`

`# Predictions`
`pred_test = trained_model.predict(X_test)`
`pred_test = [np.argmax(x) for x in pred_test]`

`# Actual`
`true_test = final_Y_test`
`true_test = [np.argmax(x) for x in true_test]`

`# Find accuracies`
`accuracy = accuracy_score(true_test, pred_test)`

`print("The total accuracy is : ", accuracy)`

I'm new to TensorFlow, if anyone has any suggestions, please raise, cheers !

**Comment by jvishnuvardhan** [comment]
@AchiraFernando It is strange. The saved model should give same performance as original. 
Could you please check the tutorial listed [here](https://github.com/tensorflow/docs/blob/r1.12/site/en/r2/tutorials/keras/save_and_restore_models.ipynb) on saving and restoring models. Please let me know how it progresses. Thanks!

**Comment by AchiraFernando** [comment]
@jvishnuvardhan Thanks for checking!
================================
I think I found the issue.
The model saving and loading works alright as expected.
But the problem is with the tokenizing of the text data and converting the labels to categorical.

1) I'm using the same tokenizer instance from the training setup works, but I’ve tried using pickle to save the tokenizer instance but that didn’t do any change. Code is as below for saving and loading. Or do you have any suggestions for saving the tokenizer ?

```
# saving
with open(‘tokenizer.pickle’, ‘wb’) as handle:
pickle.dump(tokenizer_train, handle, protocol=pickle.HIGHEST_PROTOCOL)

# load tokenizer
tokenizer = Tokenizer()
with open(‘tokenizer.pickle’, ‘rb’) as handle:
tokenizer = pickle.load(handle)
```

2) I’m using to_categorical to convert the labels (0 to 5) to one hot vectors. I’m using the below piece of code in the training session and testing session. But when testing, doing this gives me false accuracy but using the already converted values from the training session gives me the correct accuracy. Do you think we need to save this by any means as well ?

```
test_revs = pd.read_csv(‘hotel_reviews_test.csv’) # fetching the reviews from csv to a dataframe

test_revs.loc[:, ‘rating’] = test_revs[‘rating’].apply(points_to_class)

Y_test = to_categorical(test_revs[‘rating’], 5) # converting to one hot vectors with 5 as number of classes
```

**Comment by jvishnuvardhan** [comment]
Closing this out since I understand it to be resolved, but please let me know if I'm mistaken.Thanks!
